File: backend/src/modloader/fabric.ts
import got, { Got } from "got";
import fs from "node:fs"
import { execPromise, fastdownload, verifySHA1, calculateSHA1 } from "../utils/utils.js";
import config from "../utils/config.js";
import { logger } from "../utils/logger.js";

interface ILatestLoader{
    url:string,
    stable:boolean
}

interface IServer{
    libraries:{
        name:string
    }[]
}

export class Fabric{
    minecraft: string;
    loaderVersion: string;
    got:Got
    path: string;
    constructor(minecraft:string,loaderVersion:string,path:string) {
        this.minecraft = minecraft;
        this.loaderVersion = loaderVersion;
        this.path = path
        this.got = got.extend({
            prefixUrl:"https://bmclapi2.bangbang93.com/",
            headers:{
                "User-Agent": "DeEarthX"
            }
        })
    }

    async setup():Promise<void>{
        await this.installer()
        if(config.mirror.bmclapi){
            await this.libraries()
        }
        await this.install()
        await this.wshell()
    }

    async install(){
        const javaCmd = config.javaPath || 'java';
        await execPromise(`${javaCmd} -jar fabric-installer.jar server -dir . -mcversion ${this.minecraft} -loader ${this.loaderVersion}`,{
            cwd:this.path
        }).catch(e=>console.log(e))
    }

    private async wshell(){
        const javaCmd = config.javaPath || 'java';
        const cmd = `${javaCmd} -jar fabric-server-launch.jar`
        await fs.promises.writeFile(`${this.path}/run.bat`,`@echo off\n${cmd}`) //Windows
        await fs.promises.writeFile(`${this.path}/run.sh`,`#!/bin/bash\n${cmd}`) //Linux
     }

async libraries(){
    const res = await this.got.get(`fabric-meta/v2/versions/loader/${this.minecraft}/${this.loaderVersion}/server/json`).json<IServer>()
    const _downlist: [string,string,string?][]= []
    res.libraries.forEach(e=>{
        const path = this.MTP(e.name)
        _downlist.push([`https://bmclapi2.bangbang93.com/maven/${path}`,`${this.path}/libraries/${path}`])
    })
    
    // 下载所有库文件
    await fastdownload(_downlist as any)
    
    // 验证所有下载的库文件
    if (config.mirror.bmclapi) {
        logger.info(`验证 ${_downlist.length} 个 Fabric 库文件的完整性...`)
        let verifiedCount = 0
        for (const [, filePath] of _downlist) {
            if (fs.existsSync(filePath)) {
                const hash = calculateSHA1(filePath)
                logger.debug(`${filePath}: SHA1 = ${hash}`)
                verifiedCount++
            }
        }
        logger.info(`Fabric 库文件验证完成，共验证 ${verifiedCount}/${_downlist.length} 个文件`)
    }
}

async installer(){
    let downurl = ""
    const res = await this.got.get("fabric-meta/v2/versions/installer").json<ILatestLoader[]>()
    res.forEach(e=>{
        if(e.stable){
            //downurl = `https://bmclapi2.bangbang93.com/maven/${new URL(e.url).pathname.slice(1)}`
            downurl = e.url
            return;
        }
    })
    
    // 下载 installer
    const filePath = `${this.path}/fabric-installer.jar`
    await fastdownload([downurl, filePath])
    
    // 下载完成后计算并记录 hash（用于后续验证）
    if (fs.existsSync(filePath)) {
        const hash = calculateSHA1(filePath);
        logger.debug(`Fabric installer 下载完成，SHA1: ${hash}`);
    }
}

    private MTP(string:string){
        const mjp =  string.replace(/^\[|\]$/g, '')
        const OriginalName = mjp.split("@")[0]
        const x = OriginalName.split(":")
        const _mappingType = mjp.split('@')[1];
        let mappingType = ""
        if (_mappingType){
            mappingType = _mappingType
        }else{
            mappingType = "jar"
        }
            if(x[3]){
                return `${x[0].replace(/\./g, '/')}/${x[1]}/${x[2]}/${x[1]}-${x[2]}-${x[3]}.${mappingType}`
            }else{
                return `${x[0].replace(/\./g, '/')}/${x[1]}/${x[2]}/${x[1]}-${x[2]}.${mappingType}`
        }
    }    
}

File: core/modloader/fabric.go
package modloader

import (
	"encoding/json"
	"os"
	"os/exec"
	"path/filepath"

	"resty.dev/v3"

	"deearthx/core/config"
	"deearthx/core/download"
	"deearthx/core/maven"
	"deearthx/core/util"
)

// Fabric handles Fabric mod loader installation
type Fabric struct {
	minecraft     string
	loaderVersion string
	path          string
	client        *resty.Client
}

// NewFabric creates a new Fabric handler
func NewFabric(minecraft, loaderVersion, path string) *Fabric {
	return &Fabric{
		minecraft:     minecraft,
		loaderVersion: loaderVersion,
		path:          path,
		client:        resty.New(),
	}
}

// Setup installs Fabric
func (f *Fabric) Setup(progress ProgressCallback) error {
	// Download installer
	if err := f.downloadInstaller(); err != nil {
		return err
	}

	// Download libraries if BMCLAPI enabled
	cfg := config.GetConfig()
	if cfg.Mirror.BMCLAPI {
		if err := f.downloadLibraries(); err != nil {
			util.Logger.Warn("Failed to download Fabric libraries: " + err.Error())
		}
	}

	// Run installer
	if err := f.Install(); err != nil {
		return err
	}

	// Create run scripts
	return f.createRunScripts()
}

// Installer downloads the Fabric installer
func (f *Fabric) Installer() error {
	return f.downloadInstaller()
}

// Install runs the Fabric installer
func (f *Fabric) Install() error {
	cfg := config.GetConfig()
	javaCmd := "java"
	if cfg.JavaPath != "" {
		javaCmd = cfg.JavaPath
	}

	cmd := exec.Command(javaCmd, "-jar", "fabric-installer.jar", "server", "-dir", ".", "-mcversion", f.minecraft, "-loader", f.loaderVersion)
	cmd.Dir = f.path

	output, err := cmd.CombinedOutput()
	if err != nil {
		util.Logger.Error("Fabric install failed: " + string(output))
		return err
	}

	util.Logger.Info("Fabric installation complete")
	return nil
}

func (f *Fabric) downloadInstaller() error {
	// Get latest installer URL
	installerURL := download.GetBMCLAPIPrefix() + "/fabric-meta/v2/versions/installer"

	resp, err := f.client.R().SetHeader("User-Agent", "DeEarthX").Get(installerURL)
	if err != nil {
		return err
	}

	var installers []struct {
		URL    string `json:"url"`
		Stable bool   `json:"stable"`
	}

	if err := json.Unmarshal(resp.Bytes(), &installers); err != nil {
		return err
	}

	var downloadURL string
	for _, installer := range installers {
		if installer.Stable {
			downloadURL = installer.URL
			break
		}
	}

	if downloadURL == "" && len(installers) > 0 {
		downloadURL = installers[0].URL
	}

	filePath := filepath.Join(f.path, "fabric-installer.jar")

	dl := download.NewDownloadClient()
	return dl.DownloadFile(download.DownloadOptions{
		URL:        downloadURL,
		FilePath:   filePath,
		UseChunked: true,
	}, nil)
}

func (f *Fabric) downloadLibraries() error {
	// Get server JSON from fabric-meta
	url := download.GetBMCLAPIPrefix() + "/fabric-meta/v2/versions/loader/" + f.minecraft + "/" + f.loaderVersion + "/server/json"

	resp, err := f.client.R().SetHeader("User-Agent", "DeEarthX").Get(url)
	if err != nil {
		return err
	}

	var serverInfo struct {
		Libraries []struct {
			Name string `json:"name"`
		} `json:"libraries"`
	}

	if err := json.Unmarshal(resp.Bytes(), &serverInfo); err != nil {
		return err
	}

	downloadList := []download.DownloadOptions{}
	for _, lib := range serverInfo.Libraries {
		path := f.mtp(lib.Name)
		libURL := download.GetBMCLAPIPrefix() + "/maven/" + path
		libPath := filepath.Join(f.path, "libraries", path)

		downloadList = append(downloadList, download.DownloadOptions{
			URL:        libURL,
			FilePath:   libPath,
			UseChunked: true,
		})
	}

	if len(downloadList) > 0 {
		dl := download.NewDownloadClient()
		return dl.BatchDownload(downloadList, 32, nil)
	}

	return nil
}

func (f *Fabric) createRunScripts() error {
	cfg := config.GetConfig()
	javaCmd := "java"
	if cfg.JavaPath != "" {
		javaCmd = cfg.JavaPath
	}

	cmd := javaCmd + " -jar fabric-server-launch.jar"

	batContent := "@echo off\n" + cmd + "\n"
	shContent := "#!/bin/bash\n" + cmd + "\n"

	os.WriteFile(filepath.Join(f.path, "run.bat"), []byte(batContent), 0644)
	os.WriteFile(filepath.Join(f.path, "run.sh"), []byte(shContent), 0644)

	return nil
}

// mtp converts Maven coordinate to path (Fabric style)
func (f *Fabric) mtp(coord string) string {
	return maven.MTP(coord)
}